File: leetcode-submissions/198_House_Robber.ts
function rob(nums: number[]): number {
    const dp :Array<number> = new Array(nums.length).fill(-1);
    return solve(nums , 0 , nums.length, dp);
};
function solve(nums :number[], i :number , n :number , dp:Array<number>): number {
    if(i>=n){
        return 0 ;


    }
    if(dp[i]!=-1){
        return dp[i];
    }
    const take:number = nums[i]+solve(nums , i+2 , n, dp);
    const not_take:number = solve(nums , i+1, n, dp);
    return dp[i] =  Math.max(take , not_take);

}

File: leetcode-submissions/198_House_Robber.cpp
class Solution {
public:
int solve(vector<int>nums,int i,vector<int>&dp){
    if(i>=nums.size()){
        return 0;
    }
    if(dp[i]!=-1){
        return dp[i];
    }
    int left=solve(nums,i+2,dp)+nums[i];
    int right=solve(nums,i+1,dp);
    dp[i]=max(left,right);
    return dp[i];
}
    int rob(vector<int>& nums) {
       vector<int>dp(nums.size()+1,-1);
       return solve(nums,0,dp);
    }
};